lib/ansible/modules/windows/win_psmodule.ps1

```diff
@@ -90,97 +348,123 @@ Function Remove-Repository{
-      }
-      catch{
-        $ErrorMessage = "Problems installing $($Name) module: $($_.Exception.Message)"
+# Check PowerShell version, fail if < 5.0 and required modules are not installed
+$PsVersion = $PSVersionTable.PSVersion
+if ($PsVersion.Major -lt 5 ) {
+    $PrereqModulesInstalled = Install-PrereqModule -TestInstallationOnly
+    if ( -not $PrereqModulesInstalled ) {
+        $ErrorMessage = "Modules PowerShellGet and PackageManagement in versions 1.6.0 and 1.1.7 respectively have to be installed before using the win_psmodule."
         Fail-Json $result $ErrorMessage
-      }
     }
 }
 
-Function Remove-PsModule {
-    param(
-      [Parameter(Mandatory=$true)]
-      [string]$Name,
-      [bool]$CheckMode
-    )
-    # If module is present, unistalls it.
-    if (Get-Module -Listavailable|?{$_.name -eq $Name}){
-      try{
-        Uninstall-Module -Name $Name -Confirm:$false -Force -ErrorAction Stop -WhatIf:$CheckMode | out-null
-        $result.output = "Module $($Name) removed"
-        $result.changed = $true
-      }
-      catch{
-        $ErrorMessage = "Problems removing $($Name) module: $($_.Exception.Message)"
+if ( $required_version -and ( $minimum_version -or $maximum_version ) ) {
+       $ErrorMessage = "Parameters required_version and minimum/maximum_version are mutually exclusive."
+       Fail-Json $result $ErrorMessage
+}
+
+if ( $allow_prerelease -and ( $minimum_version -or $maximum_version ) ) {
+    $ErrorMessage = "Parameters minimum_version, maximum_version can't be used with the parameter allow_prerelease."
+    Fail-Json $result $ErrorMessage
+}
+
+if ( $allow_prerelease -and $state -eq "absent" ) {
+    $ErrorMessage = "The parameter allow_prerelease can't be used with state set to 'absent'."
+    Fail-Json $result $ErrorMessage
+}
+
+if ( ($state -eq "latest") -and
+    ( $required_version -or $minimum_version -or $maximum_version ) ) {
+        $ErrorMessage = "When the parameter state is equal 'latest' you can use any of required_version, minimum_version, maximum_version."
         Fail-Json $result $ErrorMessage
-      }
+}
 
+if ( $repo -and (-not $url) ) {
+    $RepositoryExists = Get-PSRepository -Name $repo -ErrorAction Ignore
+    if ( $null -eq $RepositoryExists) {
+        $ErrorMessage = "The repository $repo doesn't exist."
+        Fail-Json $result $ErrorMessage
     }
-    else{
-      $result.output = "Module $($Name) not present"
-    }
+
 }
 
-# Check powershell version, fail if < 5.0
-$PsVersion = $PSVersionTable.PSVersion
-if ($PsVersion.Major -lt 5){
-  $ErrorMessage = "Powershell 5.0 or higher is needed"
-  Fail-Json $result $ErrorMessage
+if ( ($allow_clobber -or $allow_prerelease -or $skip_publisher_check -or
+    $required_version -or $minimum_version -or $maximum_version) ) {
+    # Update the PowerShellGet and PackageManagement modules.
+    # It's required to support AllowClobber, AllowPrerelease parameters.
+    Install-PrereqModule -CheckMode $check_mode
```

Comment:
I have mixed feelings about this module updating the PSGet and PackageManagement modules instead of the user. May be good to bring this up in the next Windows Working Group.

Reply:
Updating of PSGet and PackageManagement is required to correctly perform check queries required by the module logic/purpose.

We can add an additional parameter e.g. install_prerequisites set to false.

The same issue affects the win_psrepository module - [I proposed the #PR to address that](https://github.com/ansible/ansible/pull/50759). Without updating the NuGet package provider the primary function of the module fails.